src/java/org/apache/cassandra/db/DeletionTime.java

```diff
@@ -195,42 +195,100 @@ public static Serializer getSerializer(Version version)
             return legacySerializer;
     }
 
-    // Serializer for Usigned Integer ldt
+    /* Serializer for Usigned Integer ldt
+     *
+     * ldt is encoded as a uint in seconds since unix epoch, it can go up o 2106-02-07T06:28:13+00:00 only. 
+     * Since mfda is micros since the Unix Epoch with 7 bytes we can encode 2^56 ~= 1142 years. That leaves 
+     * 1 free byte we can use to store flags for sentinel values with some space saving.
+     *
+     * Currently only a flag for the LIVE ldt is being used.
+     */
     public static class Serializer implements ISerializer<DeletionTime>
     {
+        private final static int IS_LIVE_DELETION = 0x01;
+        // Long.MAX_VALUE sentinel value equivalent in 7bytes
+        public final static long MAX_MFDA = 0xffffffffffffffL;
+
         public void serialize(DeletionTime delTime, DataOutputPlus out) throws IOException
         {
-            out.writeInt(delTime.localDeletionTimeUnsignedInteger);
-            out.writeLong(delTime.markedForDeleteAt());
+            if (delTime.equals(LIVE))
+                out.writeByte(IS_LIVE_DELETION);
+            else
+            {
+                if (delTime.markedForDeleteAt() > MAX_MFDA && delTime.markedForDeleteAt() != Long.MAX_VALUE)
+                    throw new IOException("Value too high for markForDeleteAt to encode in 7 bytes: " + delTime.markedForDeleteAt());
```

Comment:
`markedForDeleteAt` is the timestamp of a deletion. I see it as _data_. So I want to reject invalid _data_ and I don't want to silently modify _data_. Also sbdy would have to explain me why we are getting timestamps 2K years in the future and past year 2106, which is a hard limit atm bc of localDeletionTime max value. Makes sense?

Reply:
I'm wondering if it should be verified earlier - that is, when the user provides a mutation timestamp. There is also a guardrail for that: `maximum_timestamp_fail_threshold` - maybe it deserves a default value?

Reply:
Even if we verify it earlier it doesn't solve all our problems.
`MAX_MFDA` is not an implementation detail of serialization but, whether we like it or not, an externally visible limitation imposed on client-provided data.

If we start throwing, I'm worried that after the upgrade suddenly e.g. compactions start failing (because there are writes with too big MFDA already).

Thus I was wondering if stating the limitation as "all MFDA greater than MAX_MFDA will be converted to MAX_MFDA" wouldn't be a better choice. (IIUC currently it is "all MFDA greater than MAX_MFDA will break your at times, and you'd better overwrite it with smaller values")

Reply:
^Icwym, But a MFDA 2000 years into the future shouldn't happen and we should never hit that throws. We should never have an MFDA for the year 4200. The current codebase is dead in year 2106 already.

Let me ping @driftx just to double check. Brandon apologies for the ping, not trying to loop you into the review, but do you think it's reasonable to never expect a MFDA that high. Or can you think of any scenarios where this could be a problem?

Reply:
I guess a customer issuing a DELETE using a timestamp that high is the only possibility :thinking: We'd have to add a note and either fail or change the value as Kuba suggests...

Reply:
I think 2000 years is plenty reasonable.  Anyone issuing a delete that far into the future is almost certainly making a mistake...giving maximum_timestamp_fail_threshold a default value makes sense too I think, but we should float that on the ML.

Reply:
Yeah you're right. That's my exact doubt. Whether to consider that a bug/mistake and correct it rejecting it or to be more accommodating with a cap. ML is the place +1

Reply:
Email sent

Reply:
If I have gathered all opinions correctly despite the ML email getting little response:
- I added the guardrail default + a test
- I've left it to reject

I still have torn thoughts about either rejecting or capping... :thinking:

Reply:
We have another option here. Negative timestamps are not valid, so we could encode `LIVE` as a `0xFF` byte (leaving any other value above `0x7F` for other tasks later) and still correctly store any previously-valid timestamp up to `Long.MAX_VALUE`.

Because the only thing we really want to do here is save 11 bytes for live data, this would still solve the problem but will avoid having to make these decisions.

Alternatively, we could save one byte always by using the sign of the 7th byte for the `LIVE` marker, reducing the permitted values to 1000 years.

Reply:
Right lets not use the full byte until we have a justification for that. We'll need an sstable format change when we get there anyway and we might have better ideas when we revisit the ldt serialization given the recent suggestions.

I have rebase and pushed, let me know if iiuc. Thx.